src/main/java/com/gregtechceu/gtceu/integration/GTRecipeWidget.java

```diff
@@ -194,34 +192,148 @@ public GTRecipeWidget(GTRecipe recipe) {
+                    Component.translatable("gtceu.oc.tooltip.0", GTValues.VNF[getMinTier()]),
+                    Component.translatable("gtceu.oc.tooltip.1"),
+                    Component.translatable("gtceu.oc.tooltip.2"),
+                    Component.translatable("gtceu.oc.tooltip.3")
+                ));
+            addWidget(this.voltageTextWidget = voltageTextWidget);
+        }
+    }
+
+    @NotNull
+    private static List<Component> getRecipeParaText(int duration, long inputEUt, long outputEUt) {
+        List<Component> texts = new ArrayList<>();
+        texts.add(Component.literal(LocalizationUtils.format("gtceu.recipe.duration", duration / 20f)));
+        if (inputEUt != 0) {
+            texts.add(Component.literal(LocalizationUtils.format("gtceu.recipe.eu", inputEUt)));
+            texts.add(Component.literal(LocalizationUtils.format("gtceu.recipe.total", (inputEUt * duration))));
+        }
+        if (outputEUt != 0) {
+            texts.add(Component.literal(LocalizationUtils.format("gtceu.recipe.eu_inverted", outputEUt)));
+            texts.add(Component.literal(LocalizationUtils.format("gtceu.recipe.total", (outputEUt * duration))));
+        }
+        return texts;
+    }
+
+    private void addButtons() {
+        // add a recipe id getter, btw all the things can only click within the WidgetGroup while using EMI
+        int x = getSize().width + 3 - this.xOffset, y = 3;
+        if (LDLib.isEmiLoaded()) {
+            x = getSize().width - xOffset - 18;
+            y = getSize().height - 30;
+        }
+        addWidget(new PredicatedButtonWidget(x, y, 15, 15, new GuiTextureGroup(GuiTextures.BUTTON, new TextTexture("ID")), cd ->
+            Minecraft.getInstance().keyboardHandler.setClipboard(recipe.id.toString()), () -> CompassManager.INSTANCE.devMode, CompassManager.INSTANCE.devMode).setHoverTooltips("click to copy: " + recipe.id));
+    }
+
+    private int getVoltageXOffset() {
+        int x = getSize().width - switch (tier) {
+            case ULV, LuV, ZPM, UHV, UEV, UXV -> 20;
+            case OpV, MAX -> 22;
+            case UIV -> 18;
+            case IV -> 12;
+            default -> 14;
+        };
+        if (!LDLib.isEmiLoaded()) {
+            x -= 3;
+        }
+        return x;
+    }
+
+    public void setRecipeOC(int button) {
+        if (button == GLFW.GLFW_MOUSE_BUTTON_LEFT) {
+            setTier(tier + 1);
+        } else if (button == GLFW.GLFW_MOUSE_BUTTON_RIGHT) {
+            setTier(tier - 1);
+        } else if (button == GLFW.GLFW_MOUSE_BUTTON_MIDDLE) {
+            setTierToMin();
+        }
+        setRecipeTextWidget();
+    }
+
+    private void setRecipeTextWidget() {
+        long inputEUt = RecipeHelper.getInputEUt(recipe);
+        int duration = recipe.duration;
+        String tierText = GTValues.VNF[tier];
+        if (tier > getMinTier() && inputEUt != 0) {
+            LongIntPair pair = OverclockingLogic.NON_PERFECT_OVERCLOCK.getLogic().runOverclockingLogic(
```

Comment:
Why is this hardcoded to non-perfect OC? make it changeable via (maybe the UI? Maybe the recipe type?)

Reply:
oh forgot